source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -271,8 +269,11 @@ Network::IoResult TsiSocket::repeatProtectAndWrite(Buffer::Instance& buffer, boo
 
   while (true) {
     uint64_t bytes_to_drain_this_iteration =
-        prev_bytes_to_drain_ > 0 ? prev_bytes_to_drain_
-                                 : std::min(buffer.length(), max_unprotected_frame_size_);
+        prev_handshake_bytes_to_drain_ > 0
+            ? prev_handshake_bytes_to_drain_
+            : (prev_bytes_to_drain_ > 0
+                   ? prev_bytes_to_drain_
+                   : std::min(buffer.length(), actual_frame_size_to_use_ - frame_overhead_size_));
```

Comment:
This expression seems very complex.  Consider the following instead:
- ASSERT(prev_handshake_bytes_to_drain_ == 0) when entering repeatProtectAndWrite
- in TsiSocket::doWrite, check if prev_handshake_bytes_to_drain_ > 0 and if so attempt to flush the buffer and only call repeatProtectAndWrite if the flush was successful.

Reply:
Nice suggestion. Done.